app/vmctl/remoteread/remoteread.go

```diff
@@ -0,0 +1,263 @@
+	}
+	if cfg.ReadTimeout == 0 {
+		cfg.ReadTimeout = defaultReadTimeout
+	}
+
+	c := &Client{
+		c: &http.Client{
+			Timeout:   cfg.ReadTimeout,
+			Transport: http.DefaultTransport.(*http.Transport).Clone(),
+		},
+		addr:     strings.TrimSuffix(cfg.Addr, "/"),
+		user:     cfg.Username,
+		password: cfg.Password,
+	}
+	return c, nil
+}
+
+// Read fetch data from remote read source
+func (c *Client) Read(ctx context.Context, filter *Filter, streamCb StreamCallback) error {
+	query, err := c.query(filter)
+	if err != nil {
+		return fmt.Errorf("error prepare stream query: %w", err)
+	}
+	req := &prompb.ReadRequest{
+		Queries: []*prompb.Query{
+			query,
+		},
+		AcceptedResponseTypes: []prompb.ReadRequest_ResponseType{prompb.ReadRequest_STREAMED_XOR_CHUNKS},
+	}
+	data, err := proto.Marshal(req)
+	if err != nil {
+		return fmt.Errorf("unable to marshal read request: %w", err)
+	}
+
+	const attempts = 5
+	b := snappy.Encode(nil, data)
+	for i := 0; i < attempts; i++ {
+		err := c.fetch(ctx, b, streamCb)
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return fmt.Errorf("process stoped")
+			}
+			logger.Errorf("attempt %d to fetch data from remote storage: %s", i+1, err)
+			// sleeping to avoid remote db hammering
+			time.Sleep(time.Second)
+			continue
+		}
+		return nil
+	}
+	return nil
+}
+
+// Ping checks the health of the read source
+func (c *Client) Ping() error {
+	url := c.addr + healthPath
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return fmt.Errorf("cannot create request to %q: %s", url, err)
+	}
+	if c.user != "" {
+		req.SetBasicAuth(c.user, c.password)
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return fmt.Errorf("bad status code: %d", resp.StatusCode)
+	}
+	return nil
+}
+
+func (c *Client) fetch(ctx context.Context, data []byte, streamCb StreamCallback) error {
+	r := bytes.NewReader(data)
+	url := c.addr + remoteReadPath
+	req, err := http.NewRequest("POST", url, r)
+	if err != nil {
+		return fmt.Errorf("failed to create new HTTP request: %w", err)
+	}
+
+	req.Header.Add("Content-Encoding", "snappy")
+	req.Header.Add("Accept-Encoding", "snappy")
+	req.Header.Set("Content-Type", "application/x-protobuf")
+	req.Header.Set("X-Prometheus-Remote-Read-Version", "0.1.0")
+
+	if c.user != "" {
+		req.SetBasicAuth(c.user, c.password)
+	}
+
+	resp, err := c.c.Do(req.WithContext(ctx))
+	if err != nil {
+		return fmt.Errorf("error while sending request to %s: %w; Data len %d(%d)",
+			req.URL.Redacted(), err, len(data), r.Size())
+	}
+	defer func() { _ = resp.Body.Close() }()
+	if resp.StatusCode != http.StatusNoContent && resp.StatusCode != http.StatusOK {
+		body, _ := io.ReadAll(resp.Body)
+		return fmt.Errorf("unexpected response code %d for %s. Response body %q",
+			resp.StatusCode, req.URL.Redacted(), body)
+	}
+	defer func() { _ = resp.Body.Close() }()
+
+	bb := bodyBufferPool.Get()
```

Comment:
I don't think, that bufferPool is needed, use bytes.Buffer, it should be enough.

Reply:
used bytes.Buffer